src/bin/cli.rs

```diff
@@ -0,0 +1,637 @@
+            }
+
+            let ops = ops::TestOptions {
+                no_run: args.is_present("no-run"),
+                no_fail_fast: args.is_present("no-fail-fast"),
+                only_doc: doc,
+                compile_opts,
+            };
+
+            // TESTNAME is actually an argument of the test binary, but it's
+            // important so we explicitly mention it and reconfigure
+            let mut test_args = vec![];
+            test_args.extend(args.value_of("TESTNAME").into_iter().map(|s| s.to_string()));
+            test_args.extend(args.values_of("args").unwrap_or_default().map(|s| s.to_string()));
+
+            let err = ops::run_tests(&ws, &ops, &test_args)?;
+            return match err {
+                None => Ok(()),
+                Some(err) => {
+                    Err(match err.exit.as_ref().and_then(|e| e.code()) {
+                        Some(i) => CliError::new(format_err!("{}", err.hint(&ws)), i),
+                        None => CliError::new(err.into(), 101),
+                    })
+                }
+            };
+        }
+        ("uninstall", Some(args)) => {
+            let root = args.value_of("root");
+            let specs = args.values_of("spec").unwrap_or_default().collect();
+            ops::uninstall(root, specs, &values(args, "bin"), config)?;
+            Ok(())
+        }
+        ("update", Some(args)) => {
+            let ws = args.workspace(config)?;
+
+            let update_opts = ops::UpdateOptions {
+                aggressive: args.is_present("aggressive"),
+                precise: args.value_of("precise"),
+                to_update: values(args, "package"),
+                config,
+            };
+            ops::update_lockfile(&ws, &update_opts)?;
+            Ok(())
+        }
+        ("verify-project", Some(args)) => {
+            fn fail(reason: &str, value: &str) -> ! {
+                let mut h = HashMap::new();
+                h.insert(reason.to_string(), value.to_string());
+                cargo::print_json(&h);
+                process::exit(1)
+            }
+
+            let mut contents = String::new();
+            let filename = match args.root_manifest(config) {
+                Ok(filename) => filename,
+                Err(e) => fail("invalid", &e.to_string()),
+            };
+
+            let file = File::open(&filename);
+            match file.and_then(|mut f| f.read_to_string(&mut contents)) {
+                Ok(_) => {}
+                Err(e) => fail("invalid", &format!("error reading file: {}", e))
+            };
+            if contents.parse::<toml::Value>().is_err() {
+                fail("invalid", "invalid-format");
+            }
+
+            let mut h = HashMap::new();
+            h.insert("success".to_string(), "true".to_string());
+            cargo::print_json(&h);
+            Ok(())
+        }
+        ("version", _) => {
+            println!("{}", cargo::version());
```

Comment:
We may want to handle `cargo version --verbose` here as well I think?

Reply:
(although Cargo doesn't handle it today so we can deal with that later)

Reply:
Will write mentoring instructions on https://github.com/rust-lang/cargo/issues/3613 once done with this PR :)